### PR TITLE
Feature/5667 add associated records

### PR DESCRIPTION
**Pull request by HavierD**

### CODE REVIEW

File: indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+        var siblingLinks = buildLinks(siblingRecordMaps, RelationType.SIBLING);
+        if (!siblingLinks.isEmpty()) {
+            records.addAll(siblingLinks);
+        }
+
+        var childLinks = buildLinks(childRecordMaps, RelationType.CHILD);
+        if (!childLinks.isEmpty()) {
+            records.addAll(childLinks);
+        }
```
Comment (vietnguyengit) [inline indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java:37]: should consider renaming this block, something be like:

```java
var siblingRecordsLinks = getLinksFromMultiRecordsByRelationType(siblingRecords, RelationType.SIBLING);
if (!siblingRecordsLinks.isEmpty()) {
    records.addAll(siblingRecordsLinks);
}

var childRecordsLinks = getLinksFromMultiRecordsByRelationType(childRecords, RelationType.CHILD);
if (!childRecordsLinks.isEmpty()) {
    records.addAll(childRecordsLinks);
}
```

File: indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        var parentsRecordMap = getRecordMaps(data, "parent");
+        if (!parentsRecordMap.isEmpty()) {
+            var parentLink = buildLink(parentsRecordMap.get(0), RelationType.PARENT);
+            if (parentLink != null) {
+                records.add(parentLink);
+            }
+        }
+        var siblingRecordMaps = getRecordMaps(data, "brothersAndSisters");
+        var childRecordMaps = getRecordMaps(data, "children");
```
Comment (vietnguyengit) [inline indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java:27]:
```java
var siblingRecords = getRecordsByRelation(data, "brothersAndSisters");
var childRecordss = getRecordsByRelation(data, "children");
```

**Review by vietnguyengit — CHANGES_REQUESTED** [review]
let's address these naming convention feedback first and I look further later.

**Comment by vietnguyengit** [comment]
at the end, I'm confused when reading the code, let's fix and make it clearer first before analysing the functionality aspect of this PR.

File: indexer/src/main/java/au/org/aodn/esindexer/model/RelationType.java [hunk 3]
```diff
@@ -5,6 +5,9 @@
 @Getter
 public enum RelationType {
     SELF("self"),
+    PARENT("parent"),
+    SIBLING("sibling"),
+    CHILD("child"),
```
Comment (vietnguyengit) [inline indexer/src/main/java/au/org/aodn/esindexer/model/RelationType.java:10]: why not adding "brothersAndSisters" to here? so you can consistently use `RelationType` enum with your `getRecordMaps` currently uses "String key" (getRecordMaps still need to be renamed as commented)

File: indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+public class AssociateRecordFactory {
+
+    private record RecordTitle(String recordTitle, String recordAbstract) {}
+
+    public static List<LinkModel> buildAssociatedRecords(Map<String, ?> data) {
+        var records = new ArrayList<LinkModel>();
+        if (data == null || data.isEmpty()) {
+            return records;
+        }
+        var parentsRecordMap = getRecordMaps(data, "parent");
```
[the remark this answers is not in the document]
Comment (HavierD) [reply indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java:19]: Thank you. At the beginning all them are plural, but then i made parent singular but forgot synchronizing the variable names. `XXXMaps` are also the same. Previously both `XXXRecords` and `XXXRecordMaps` are existing. One is record and another is a raw data(map format) of the record. Then i simplified the code and didn't change the names. Will modify shortly

File: indexer/src/main/java/au/org/aodn/esindexer/model/RelationType.java (hunk 3 above)
Comment (HavierD) [reply indexer/src/main/java/au/org/aodn/esindexer/model/RelationType.java:10]: The `RelationType` enum contains recommended values from the [stac collection ](https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md#relation-types), where as the "brotherAndSisters" (also including "children") are `key` in a `map` from the geonetwork api. "parent" is the same, but "children" and "brothersAndSisters" are not good values of Relation Type in stac collection specification. I don't want to confuse others (someone may use "brothersAndSisters" in `rel` field of stac collection, which is bad). So i didn't put them all into the `RelationType` enum, and just use `key` for the relation of geonetwork api, to differentiate with the `RelationType` in stac collection.

File: indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                    .href(href)
+                    .rel(relationType.getValue())
+                    .title(JsonUtil.toJsonString(recordTitle))
+                    .type(MediaType.APPLICATION_JSON.getValue())
+                    .build();
+        }).orElse(null);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static List<Map<String, Object>> getRecordMaps(Map<String, ?> associatedRecordMap, String key) {
```
[the remark this answers is not in the document]
Comment (HavierD) [reply indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java:77]: Thank you! I may prefer `getRecordsByKey`. as just mentioned, I don't want to confuse relation type from stac-collection, and the `key` from geonetwork api.

File: indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java (hunk 1 above)
Comment (HavierD) [reply indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java:37]: i refactored this part to be clearer

File: indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (vietnguyengit) [reply indexer/src/main/java/au/org/aodn/esindexer/model/AssociateRecordFactory.java]: Ok maybe clearer for what "key" is, relation key?

File: indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import au.org.aodn.esindexer.model.MediaType;
+import au.org.aodn.esindexer.model.RelationType;
+import au.org.aodn.stac.model.LinkModel;
+import au.org.aodn.stac.util.JsonUtil;
+
+import java.util.*;
+
+import static au.org.aodn.esindexer.utils.CommonUtils.safeGet;
+
+public class AssociateRecordsUtil {
```
Comment (vietnguyengit) [inline indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java:12]: Associate**d** - adj ed

File: indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    private record TitleWithAbstract(String recordTitle, String recordAbstract) {}
+
+    public static List<LinkModel> generateAssociatedRecords(Map<String, ?> data) {
+        var records = new ArrayList<LinkModel>();
+        if (data == null || data.isEmpty()) {
+            return records;
+        }
+        var parentsRecord = getParentRecord(data);
+        if (!parentsRecord.isEmpty()) {
+            var parentLink = buildLink(parentsRecord, RelationType.PARENT);
```
Comment (vietnguyengit) [inline indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java:23]: parentsRecord --> parentRecord (parent is adj)

File: indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    private record TitleWithAbstract(String recordTitle, String recordAbstract) {}
+
+    public static List<LinkModel> generateAssociatedRecords(Map<String, ?> data) {
+        var records = new ArrayList<LinkModel>();
+        if (data == null || data.isEmpty()) {
+            return records;
+        }
+        var parentsRecord = getParentRecord(data);
+        if (!parentsRecord.isEmpty()) {
```
Comment (vietnguyengit) [inline indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java:22]: no "s" after "parent"

File: indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        return Collections.emptyMap();
+    }
+
+    private static List<Map<String, Object>> getSiblingRecords(Map<String, ?> associatedRecordMap) {
+        return getRecords(associatedRecordMap, "brothersAndSisters");
+    }
+
+    private static List<Map<String, Object>> getChildRecords(Map<String, ?> associatedRecordMap) {
+        return getRecords(associatedRecordMap, "children");
+    }
```
Comment (vietnguyengit) [inline indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java:89]: why would you break it down like this? just call getRecordsByRelation and you solve the problem.

**Review by vietnguyengit — CHANGES_REQUESTED** [review]

File: indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java (hunk 8 above)
Comment (vietnguyengit) [reply indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java:22]: I didn't check all every single line, but please double check adjective and plural usages

File: indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java (hunk 9 above)
Comment (HavierD) [reply indexer/src/main/java/au/org/aodn/esindexer/utils/AssociateRecordsUtil.java:89]: I use a recommended style, if function argument contains flag / very few possible values, try to break this function into several functions. This function is simple, maybe the benefits of "breaking into several functions" is not that much and obvious.

File: indexer/src/main/java/au/org/aodn/esindexer/utils/AssociatedRecordsUtil.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import au.org.aodn.esindexer.model.MediaType;
+import au.org.aodn.esindexer.model.RelationType;
+import au.org.aodn.stac.model.LinkModel;
+import au.org.aodn.stac.util.JsonUtil;
+
+import java.util.*;
+
+import static au.org.aodn.esindexer.utils.CommonUtils.safeGet;
+
+public class AssociatedRecordsUtil {
```
Comment (vietnguyengit) [inline indexer/src/main/java/au/org/aodn/esindexer/utils/AssociatedRecordsUtil.java:12]: a lot easier to read now

File: indexer/src/main/java/au/org/aodn/esindexer/utils/CommonUtils.java [hunk 11]
```diff
@@ -15,4 +16,26 @@ public static <T> Optional<T> safeGet(Supplier<T> supplier) {
...
         }
     }
+
+    // alternative function for @Retryable annotation when the class is not a spring bean
+    public static void persevere(BooleanSupplier action) {
+        persevere(10, 1, action);
+    }
+
+
+
```
Comment (vietnguyengit) [inline indexer/src/main/java/au/org/aodn/esindexer/utils/CommonUtils.java:26]: excessive line breaks

File: indexer/src/main/java/au/org/aodn/esindexer/utils/AssociatedRecordsUtil.java (hunk 10 above)
Comment (HavierD) [reply indexer/src/main/java/au/org/aodn/esindexer/utils/AssociatedRecordsUtil.java:12]: thank you. will remove soon

File: indexer/src/main/java/au/org/aodn/esindexer/utils/CommonUtils.java (hunk 11 above)
Comment (HavierD) [reply indexer/src/main/java/au/org/aodn/esindexer/utils/CommonUtils.java:26]: thank you. will remove soon

**Review by vietnguyengit — APPROVED** [review]

**Pull request merged**